crates/wasm-smith/src/core.rs

```diff
@@ -549,16 +569,20 @@ impl Module {
             let max_rec_group_size = self.config.max_types - self.types.len();
             let rec_group_size = u.int_in_range(0..=max_rec_group_size)?;
             let type_ref_limit = u32::try_from(self.types.len() + rec_group_size).unwrap();
+            self.max_type_limit = MaxTypeLimit::Num(type_ref_limit);
             for _ in 0..rec_group_size {
-                let ty = self.arbitrary_sub_type(u, type_ref_limit)?;
+                let ty = self.arbitrary_sub_type(u)?;
                 self.add_type(ty);
             }
         } else {
             let type_ref_limit = u32::try_from(self.types.len()).unwrap();
-            let ty = self.arbitrary_sub_type(u, type_ref_limit)?;
+            self.max_type_limit = MaxTypeLimit::Num(type_ref_limit);
+            let ty = self.arbitrary_sub_type(u)?;
             self.add_type(ty);
         }
 
+        self.max_type_limit = MaxTypeLimit::ModuleTypes;
```

Comment:
Might be worth it to have 

```rust
fn with_max_type_limit<T>(&mut self, limit: MaxTypeLimit, f: impl FnMut(&mut Self) -> T) -> T {
    ...
}
```

Reply:
With only one method that modifies the max type limit I think I'll stick to this for now because I never like to rename `self` to `me` or `this`, but if this grows in the future I agree a helper should be added